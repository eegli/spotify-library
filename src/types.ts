import { AxiosResponse } from 'axios';

type PickActual<T, P extends keyof T> = T[P];

export type AppConfig = {
  token: string;
  type: 'full' | 'normal';
  genres: boolean;
  features: boolean;
  outDir: string;
};

export type TrackLight = {
  id: string;
  name: string;
  album: {
    name: string;
    id: string;
  };
  artists: {
    id: string;
    name: string;
  }[];
};

export type TrackFull = PickActual<SpotifyApi.SavedTrackObject, 'track'>;

export type Library<T> = {
  added_at: string;
  track: T & {
    genres?: string[][];
    features?: SpotifyApi.AudioFeaturesObject;
  };
}[];

export type UsersSavedTracks =
  AxiosResponse<SpotifyApi.UsersSavedTracksResponse>;
export type MultipleArtists = AxiosResponse<SpotifyApi.MultipleArtistsResponse>;
export type MultipleAudioFeatures =
  AxiosResponse<SpotifyApi.MultipleAudioFeaturesResponse>;
