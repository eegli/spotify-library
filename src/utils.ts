import cliProgress from 'cli-progress';
import fs from 'fs';
import { join } from 'path';

export function hasOwnProperty<
  T extends Record<PropertyKey, unknown>,
  K extends PropertyKey
>(obj: T, prop: K): obj is T & Record<K, unknown> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

export function write(path: string, fileName: string, data: unknown): string {
  path = join(process.cwd(), path);
  if (!fs.existsSync(path)) {
    fs.mkdirSync(path, { recursive: true });
  }
  if (fileName.endsWith('.json')) {
    fileName = fileName.slice(0, -5);
  }
  path = join(path, fileName + '.json');
  fs.writeFileSync(path, JSON.stringify(data, null, 2));
  return path;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function chunkify<T extends any[] | Set<any>>(
  obj: T,
  chunkSize: number
): T[][] {
  const len = Array.isArray(obj) ? obj.length : obj.size;
  const chunks = [];
  for (let i = 0; i < len; i += chunkSize) {
    chunks.push([...obj].slice(i, i + chunkSize));
  }
  return chunks;
}

export function createProgressBar(items: string) {
  return new cliProgress.SingleBar({
    format:
      `Fetching ${items}\t |` +
      '{bar} {percentage}% | ETA: {eta}s | {value}/{total}',
    barCompleteChar: '\u25A0',
    barIncompleteChar: ' ',
    hideCursor: true,
  });
}

export function goodbye(message: string): never {
  console.error('\n', '\x1b[31m', `Error: ${message}. Goodbye`, '\x1b[0m');
  process.exit(1);
}
