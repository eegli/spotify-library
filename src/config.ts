import { AppConfig } from './types';
import { goodbye, hasOwnProperty } from './utils';

export const defaultConfig: AppConfig = {
  token: '',
  type: 'normal',
  genres: false,
  features: false,
  outDir: '',
};

export const createConfig = (args: string[]): AppConfig => {
  const config: AppConfig = { ...defaultConfig };

  args.forEach((val, idx, orig) => {
    if (val.startsWith('--')) {
      const key = val.slice(2);
      if (hasOwnProperty(config, key)) {
        // Flag argument
        if (orig[idx + 1] && !orig[idx + 1].startsWith('--')) {
          config[key] = orig[idx + 1];
          // Boolean flag
        } else {
          config[key] = true;
        }
      } else {
        console.warn(`Ignoring unknown option: ${val}`);
      }
    }
  });

  if (!['full', 'normal'].includes(config.type)) {
    config.type = defaultConfig.type;
  }

  if (!config.token) {
    goodbye('Missing Spotify authentication token');
  }

  return config;
};
