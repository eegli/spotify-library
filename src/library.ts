import axios from 'axios';
import { parseConfig } from './config';
import type {
  Library,
  MultipleArtists,
  MultipleAudioFeatures,
  TrackFull,
  TrackLight,
  UsersSavedTracks,
} from './types';
import { chunkify, createProgressBar, goodbye, write } from './utils';

export async function library(): Promise<Library<TrackLight | TrackFull>> {
  const config = await parseConfig(process.argv.slice(2));
  try {
    let library: Library<TrackFull | TrackLight> = [];
    let progress = createProgressBar('user library');

    const pagination = {
      nextUrl: <string | null>'https://api.spotify.com/v1/me/tracks',
      start: 0,
      end: 50,
    };

    progress.start(0, 0);
    while (pagination.nextUrl) {
      const { data }: UsersSavedTracks = await axios.get(pagination.nextUrl, {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
        params: {
          limit: 50,
        },
      });

      progress.setTotal(data.total);
      progress.increment(data.items.length);

      data.items.forEach(el => {
        library.push(el);
      });

      pagination.nextUrl = data.next;
      pagination.start = pagination.end;
      pagination.end += data.limit;
    }

    progress.stop();

    // Reduce library if necessary
    if (config.type === 'normal') {
      library = library.reduce((acc, curr) => {
        acc.push({
          added_at: curr.added_at,
          track: {
            id: curr.track.id,
            name: curr.track.name,
            album: {
              id: curr.track.album.id,
              name: curr.track.album.name,
            },
            artists: curr.track.artists.map(a => ({
              id: a.id,
              name: a.name,
            })),
          },
        });
        return acc;
      }, <Library<TrackLight>>[]);
    }

    // Add genres if specified
    if (config.genres) {
      const artists = library.map(t => t.track.artists.map(a => a.id)).flat();
      const artistIds = new Set<string>(artists);
      const genres: Record<string, string[]> = {};

      // Max 50 items per request to Spotify
      const chunkSize = 50;

      progress = createProgressBar('artists');
      progress.start(artistIds.size, 0);

      for await (const chunk of chunkify(artistIds, chunkSize)) {
        const { data }: MultipleArtists = await axios.get(
          'https://api.spotify.com/v1/artists',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (config.features) {
      const songIds = library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      const chunkSize = 50;
      progress = createProgressBar('audio features');

      progress.start(songIds.length, 0);
      for await (const chunk of chunkify(songIds, 50)) {
        const { data }: MultipleAudioFeatures = await axios.get(
          'https://api.spotify.com/v1/audio-features',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );

        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const outDir = write(config.outDir, 'spotify-library', library);
    console.info("Success! Library written to '%s'", outDir);
    return library;
  } catch (e) {
    goodbye('Something went wrong: ' + e);
  }
}
