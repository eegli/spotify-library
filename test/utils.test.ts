import fs from 'fs';
import { chunkify, write } from '../src/utils';

const mockFS = fs as jest.Mocked<typeof fs>;

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Utils', () => {
  it('chunkify with arr', () => {
    const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    expect(chunkify(arr, 3)).toEqual([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10]]);
  });

  it('chunkify with set', () => {
    const set = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    expect(chunkify(set, 3)).toEqual([[1, 2, 3], [4, 5, 6], [7, 8, 9], [10]]);
  });

  const writeParams: Parameters<typeof write>[] = [
    ['tmp', 'test', { data: true }],
    ['tmp/dir', 'test.json', { data: true }],
    ['/tmp/dir/', 'test', { data: true }],
    ['../dir', 'test.json', { data: true }],
  ];
  writeParams.forEach((args, idx) => {
    it(`writes data, ${idx}`, () => {
      write(...args);
      expect(mockFS.writeFileSync).toHaveBeenCalledTimes(1);
      const outDirPath = (
        mockFS.writeFileSync.mock.calls[0][0] as string
      ).replace(/\\|\//gi, '/');

      expect(outDirPath).toMatchSnapshot(`write file, ${idx}`);
    });
  });
});
