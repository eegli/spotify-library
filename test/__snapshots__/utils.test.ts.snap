// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils writes data, 0: write file, 0 1`] = `"/usr/projects/stuff/tmp/test.json"`;

exports[`Utils writes data, 1: write file, 1 1`] = `"/usr/projects/stuff/tmp/dir/test.json"`;

exports[`Utils writes data, 2: write file, 2 1`] = `"/usr/projects/stuff/tmp/dir/test.json"`;

exports[`Utils writes data, 3: write file, 3 1`] = `"/usr/projects/dir/test.json"`;
